File: Templates/Python/Clustering/HierarchicalClustering.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jan 27 13:58:36 2019

@author: Kenil Shah
"""
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

#Importing Dataset
dataset = pd.read_csv("Salary_Data.csv")

#Gives us a summary of the dataset
dataset.describe()

#to give column names
dataset.columns = []

#Diving the dataset into independent and dependent variables
X = dataset.iloc[:,:].values
Y = dataset.iloc[:,:].values

# Column names
list(dataset)

#Correlation Matrix
corr_matrix = dataset.corr()
f, ax = plt.subplots(figsize=(16,10))
ax = sns.heatmap(corr_matrix)
ax.set_title("correlation between all features")
figure = ax.get_figure()    

#Do feature scaling if required.
#Convert into categorical variable if required

#Using dendogram to find optimal number of cluster
import scipy.cluster.hierarchy as sch
dendogram = sch.dendrogram(sch.linkage(X,method = 'ward'))   # ward try to minimize the within cluster variance
plt.title('Dendogram')
plt.xlabel('Customers')
plt.ylabel('Euclidean Distance')
plt.show()

#Model
from sklearn.cluster import AgglomerativeClustering
model = AgglomerativeClustering(n_clusters = 5,affinity = 'euclidean',linkage = 'ward') # n_clusters are the number of clusters
                                                                     # affinity tells about the type of distance we want to use
pred = model.fit_predict(X)

File: Templates/R/Clustering/HierarchicalClustering.R

dataset <- read.csv('filename')
summary(dataset) # provides a summary of the dataset

str(dataset) # gives us a structure of the dataset

names(dataset) # names of the column

#check for correlation 

#tocheck for two columns

cor(dataset&column1,dataset&column2,method = 'pearson') # other methods can be used

#between all the variables

library(ggpubr)
ggscatter(dataset,x= 'column 1',y = 'column 2',add = 'reg.line',
          conf.int = TRUE , cor.coef = TRUE , cor.method = 'pearson',
          xlab = 'Column 1',ylab = 'Column 2')

#Correltion diagram

library(corrplot)
x <- cor(dataset)
corrplot(x, type="upper", order="hclust")

#Correlation Matrix

corr.test(dataset)

#Correlation of one variable will every variable

library(corrr)
dataset %>% correlate() %>% focus(columnname)

#Do scaling if needed
#Convert into categorical variable if required

#Before creating the model, you might need to bin few values . You can use mapvalue function for that
library(plyr)
dataset$column <- mapvalues(dataset$column,from = c('A','B','C'),to = c('D','E','F')) # to map A to D, B to E,c to F

#Using dendogram to find the optimal number of clusters
dendogram = hcclust(dist(X,method = 'euclidean',method = 'ward.D'))  # first argument is the dissimilarity structure 
                                                  #method ward tries to minimize the within cluser variances

plot(dendogram,main = paste('Dendogram'),xlab = 'Customers', ylab = 'Euclidean distances')

#Model
model <- hclust(dist(X,method = 'euclidean'),method = 'ward.D')
finalmodel <- cutree(model,5)










